### ISSUE TITLE
Low accuracy compared to non-Bayesian PyTorch model

### ISSUE THREAD

**Issue by JSchmie** [RAI-SCC/torch_blue#49]
Hi,

As part of the [JOSS review](https://github.com/openjournals/joss-reviews/issues/9415), I adapted one of my existing tutorials to use `torch_blue`. Afterwards, I noticed that you use a very similar approach in your own [tutorial](https://github.com/RAI-SCC/torch_blue/blob/main/scripts/pytorch_tutorial.py). Replacing the standard PyTorch layers with your `vi` layers worked smoothly, and I was able to train the model without any issues. However, the final test performance is significantly worse than for the corresponding non-Bayesian model.

<details>
<summary>For context, my current code</summary>

```python
class simpleNN(vi.VIModule): 
    def __init__(self, input_size, num_classes) -> None:
        super(simpleNN,self).__init__()
        self.fc1 = vi.VILinear(input_size,50) 
        self.fc2 = vi.VILinear(50,num_classes) 
    def forward(self,x):
        x = F.relu(self.fc1(x)) 
        x = self.fc2(x) 
        return x 


input_size = 784 
num_classes = 10 
learning_rate = 0.001 
num_epochs = 25 

model = simpleNN(input_size= input_size, num_classes= 10).to(device) 
predictive_distribution = Categorical()
loss_fn = vi.KullbackLeiblerLoss(predictive_distribution, dataset_size=total_train_images)
optimizer = optim.Adam(model.parameters(),lr=learning_rate)

def train():
    loss_arr = []
    for epoch in trange(num_epochs):
        epoch_loss = 0
        for batch_idx,(img,label) in enumerate(trainloader):

            data = img.to(device=device)
            target = label.to(device=device)
            
            data = data.reshape(data.shape[0],-1) # reshape from torch.Size([64, 1, 28, 28])  torch.Size([64, 784]) which flattens the input

            score = model(data)
            loss = loss_fn(score,target)
            epoch_loss += loss.item()

            optimizer.zero_grad() 
            loss.backward() 

            optimizer.step() 
        avg_epoch_loss = epoch_loss / len(trainloader)
        loss_arr.append(avg_epoch_loss)
        print(f"Epoch {epoch+1}/{num_epochs}, Loss: {avg_epoch_loss:.4f}")
    plt.plot(loss_arr)

train()

def check_accuracy(loader,model):
    size= len(loader.dataset)
    num_batches = len(loader)
    num_correct = 0.0
    test_loss = 0.0
    model.eval()
    
    with torch.no_grad():
        for x,y in loader:
            x = x.to(device = device)
            y = y.to(device = device)
            
            x = x.reshape(x.shape[0],-1)
            
            scores = model(x)

            test_loss += loss_fn(scores, y).item()
            
            predictions = predictive_distribution.predictive_parameters_from_samples(
                    scores[0]
                )
            
            num_correct += (predictions.argmax(0) == y).type(torch.float).sum().item()
    num_correct /= size
    test_loss /= num_batches    
    print(
        f"Test Error: \n Accuracy: {(100 * num_correct):>0.1f}%, Avg loss: {test_loss:>8f} \n"
    )
    model.train()

check_accuracy(trainloader, model)
check_accuracy(testloader, model)
```

</details>

`scores` had the size `torch.Size([10, 64, 10])`

The Accuracy for the test set is 14.8% using torch blue vs 86.76% using the non-Bayesian model. 
Could you tell me what I did wrong or why the accuracy is so low?

**Comment by ArvidWeyrauch** [comment]
Hi,

thank you for trying things. Unfortunately, it is likely the answer to your question is that VI BNNs can be a little tricky to get right. I am setting up a debugging guide in the documentation that gives some common ways to approach this, which can be found [here](https://github.com/RAI-SCC/torch_blue/blob/docs_patch/docs/source/debug.rst) (still on a different branch).

Most importantly, all torch_blue losses have the `track` argument, which can track the terms of the loss separately. Plotting the components (and comparing them) can often help pinpoint what is going wrong. `data_fitting` behaves roughly like a non-Bayesian loss. If `prior_matching` is too dominant it might induce behavior that looks like overfitting. (If you post plots here I might be able to provide further insight)

As a short summary, there are three common things to do:
1. Setting the `heat` parameter to a number smaller than 1, which scales back the prior matching term. This is related to the cold posterior effect and may degrade calibration if done too aggressively. However, it sometimes stops the prior term from blurring away the training progress.
2.  Changing the initial std of the variational distribution, which can stabilize the early stages of training (both lowering or raising might be necessary, this would be a longer discussion). Learning rate warm up might also help here.
3. Changing (typically lowering) the prior std, as it might pull the weights to unreasonably large stds.

I am working on adding tools and guidance to help with this, so I am happy for any feedback on what users have trouble with and also if you find new solutions.

**Comment by JSchmie** [comment]
Thank you; I set it now and got an accuracy of 87.3% with predictions.argmax(-1). 
A suggestion: as part of your ReadTheDocs, you could include a demo Jupyter file with explanations.
